java-manta-client/src/main/java/com/joyent/manta/client/multipart/EncryptedMultipartManager.java

```diff
@@ -0,0 +1,363 @@
+         * encrypted it and stored it in m-encrypt-metadata. */
+        metadata.removeAllEncrypted();
+
+        /* If content-length is specified, we store the content length as
+         * the plaintext content length on the object's metadata and then
+         * remove the header because server-side MPU will attempt to verify
+         * that the content-length in the header is the same as the ciphertext's
+         * length. This won't work because the ciphertext length will always
+         * be different than the plaintext content length. */
+        if (headers.getContentLength() != null && headers.getContentLength() > -1) {
+            metadata.put(MantaHttpHeaders.ENCRYPTION_PLAINTEXT_CONTENT_LENGTH,
+                    headers.getContentLength().toString());
+            headers.remove(HttpHeaders.CONTENT_LENGTH);
+        } else if (contentLength != null && contentLength > -1) {
+            metadata.put(MantaHttpHeaders.ENCRYPTION_PLAINTEXT_CONTENT_LENGTH,
+                    contentLength.toString());
+        }
+
+        /* If the Content-MD5 header is set, it will always be inaccurate because
+         * the ciphertext will have a different checksum and it will cause the
+         * server-side checksum verification to fail. */
+        if (headers.getContentMD5() != null) {
+            headers.remove(HttpHeaders.CONTENT_MD5);
+        }
+
+        WRAPPED_UPLOAD upload = wrapped.initiateUpload(path, metadata, headers);
+
+        EncryptionState encryptionState = new EncryptionState(encryptionContext);
+        EncryptedMultipartUpload<WRAPPED_UPLOAD> encryptedUpload =
+                new EncryptedMultipartUpload<>(upload, encryptionState);
+
+        LOGGER.debug("Created new encrypted multipart upload: {}", upload);
+
+        return encryptedUpload;
+    }
+
+    @Override
+    public Stream<MantaMultipartUpload> listInProgress() throws IOException {
+        return wrapped.listInProgress();
+    }
+
+    @Override
+    protected MantaMultipartUploadPart uploadPart(final EncryptedMultipartUpload<WRAPPED_UPLOAD> upload,
+                                                  final int partNumber,
+                                                  final HttpEntity sourceEntity) throws IOException {
+        Validate.notNull(upload, "Multipart upload object must not be null");
+
+        if (!upload.canUpload()) {
+            String msg = "Multipart object is not in a state that it can be "
+                    + "used for uploading parts. For encrypted multipart "
+                    + "uploads, only multipart upload objects returned from "
+                    + "the initiateUpload() method can be used to upload parts.";
+            throw new IllegalArgumentException(msg);
+        }
+
+        final EncryptionState encryptionState = upload.getEncryptionState();
+        encryptionState.getLock().lock();
+
+        final EncryptionContext encryptionContext = encryptionState.getEncryptionContext();
```

Comment:
I don't know how this could fail, but as  a matter of hygiene the log should be the last thing before the t`ry`.